Core/Chat: general cleanup and revamping:

ariel- opened the pull request:
**Changes proposed**: General cleanup of code in game/Chat folder
- Const correctness
- Code standards
- Proper use of map iterators and map::at instead of operator[]-ing-everything

**Target branch(es)**: 335/6x

**Tests performed**: (Does it build, tested in-game, etc)
It builds and works properly.

Aokromes commented:
@ariel- can you give a look to #8411 ? :)

ariel- commented:
Interesting, will take a look for sure 👍

Shauren commented on src/server/game/Chat/Channels/Channel.h:
```diff
@@ -269,19 +284,21 @@ class TC_GAME_API Channel
-        BannedContainer bannedStore;
+        typedef GuidUnorderedSet BannedContainer;
+
+        bool AnnounceEnabled;           //< Whether we should broadcast a packet whenever a player joins/exits the channel
```
all of these names should start with underscore and lowercase letter if they are not public

Shauren commented on src/server/game/Chat/Channels/Channel.cpp:
```diff
@@ -288,7 +306,8 @@ void Channel::KickOrBan(Player const* player, std::string const& badname, bool b
     }
 
-    if (!playersStore[good].IsModerator() && !player->GetSession()->HasPermission(rbac::RBAC_PERM_CHANGE_CHANNEL_NOT_MODERATOR))
+    PlayerInfo& info = PlayersStore.at(good);
```
is the `good` key guaranteed to be found in the map at this point?

ariel- replied:
if (!IsOn(good)) check handles this case and stops execution prematurely before reaching this line.

jackpoz commented on src/server/game/Chat/Channels/Channel.cpp:
```diff
@@ -191,36 +191,38 @@ void Channel::JoinChannel(Player* player, std::string const& pass)
     MakeYouJoined(&data);
     SendToOne(&data, guid);
 
-    JoinNotify(guid);
```
by looking at this PR I have no clue if this line was removed on purpose (whatever the purpose would be) or if it was removed by accident

ariel- replied:
Silly me, I removed it when testing 💃

jackpoz commented on src/server/game/Chat/Channels/Channel.cpp:
```diff
@@ -191,36 +191,38 @@ void Channel::JoinChannel(Player* player, std::string const& pass)
         // If the channel has no owner yet and ownership is allowed, set the new owner.
-        if (!_ownerGUID && _ownership)
+        // or if the owner was a GM with .gm visible off
+        // don't do this if the new player is, too, an invis GM, unless the channel was empty
```
this ".gm visible off" stuff seems unrelated to "Changes proposed:" section of the PR description

ariel- replied:
Well, I added those changes in the commit description, and I don't feel like repeating myself, if you want I can copy them to the PR desc too

jackpoz commented:
The (misleading) description of this PR suggests that there are only codestyle changes while instead there is a mix of both codestyle and behavior changes, making it impossible to review and being a pain to revert if needed.

The codestyle changes on their own are very difficult to check, there is no way to ensure all variables got renamed without any typo to a wrong name unless they are renamed locally too using IDEs like Visual Studio. A PR with just const-correctness and another PR with just std::map operators would have allowed a much smoother review.

ariel- commented:
The renames were done with VS2015 Rename feature.

ariel- committed:
Core/Chat: general cleanup and revamping:

- Remove duplicate ObjectGuid in Channel (both as key and value member)
- Add const-ness to Channel methods
- NULL->nullptr
- const xx* -> xx const*
- Removed all instances of std::map::operator[] except where it is actually INTENDED to insert an element
- Use Trinity::StringFormat instead of snprintf

ariel- commented:
Well I'm splitting this. The PR will now handle general nice up of code, instead of changing how it works, updating....

ariel- commented:
Test fix for #8411: http://i.imgur.com/COD7MNY.png

Required a major rewrite in ChannelMgr/Channel.cpp, I have locally branched this off this PR; I'm warning there were many lines changed; So, should I close this and squash it with the other one? Or do you prefer reviewing this first?

jackpoz commented:
this PR looks ready to be merged (if those .at() changes have already been tested and are not going to trigger any issue).

jackpoz commented:
btw is ```char const*``` the TC standard ?

Shauren commented:
you mean const and * placement? yes.

Rochet2 commented on src/server/game/Chat/Channels/Channel.h:
```diff
@@ -148,106 +146,121 @@ class TC_GAME_API Channel
+        void SendToOne(WorldPacket* data, ObjectGuid who) const;
+
+        bool IsOn(ObjectGuid who) const { return _playersStore.count(who) != 0; }
+        bool IsBanned(ObjectGuid guid) const { return _bannedStore.count(guid) != 0; }
```
Isn't using count considered a bad practice? Guess it doesnt matter really, but just thought it was weird that this was changed for no real reason.

Shauren replied:
No, this is the simplest way to check if an element exists in a set

Rochet2 commented on src/server/game/Chat/Channels/Channel.h:
```diff
@@ -269,19 +286,20 @@ class TC_GAME_API Channel
 
         typedef std::map<ObjectGuid, PlayerInfo> PlayerContainer;
-        typedef GuidSet BannedContainer;
+        typedef GuidUnorderedSet BannedContainer;
```
Since GuidSet was changed to unordered, maybe PlayerContainer can be changed to unorderd as well?
Not sure if it can be just swapped. Cant see any uses in this diff for the map's ordering.

jackpoz replied:
inb4 PR to change it from unordered to ordered

ariel- replied:
Well, BannedContainer is just for element checking, so the change to unordered can't hurt.

PlayerContainer on the other hand; being ordered means that when a player leaves, next owner is searched from the player with the lowest guid in the channel, and as discussed in the PR page, this is meant to only be a cleanup, not modify the behaviour, and making PlayerContainer unordered will change it.
